### DISCUSSION TITLE
IPAM via a REST API

### DISCUSSION THREAD

**Discussion by techch01** [Azure/ipam#27] [General]
Hello,

Can we have API verbs commands (GET, PUT, etc) until the subnet level as : /api/spaces/{space}/blocks/{block}/{networks]/**subnets**

Thanks in advance.

**Comment by DCMattyG** [comment]
Hi there @techch01! Can you help me better understand the intended use case you're wanting to implement here?

GET shouldn't be a problem, but when we start talking about PUT/PATCH/POST/etc, the IPAM tool wasn't design to modify Azure resources. It's current design is essentially to read and bubble up the relevant information to the user and bring visibility to what's happening within Azure. Are you saying here you want to be able to create/u[date/delete subnets from within IPAM?

As always, thanks for the ongoing feedback and discussions!

**Reply by techch01** [reply]
Hi @DCMattyG,

Thank you for your feedback.

Yes at first I thought that IPAM allows you to modify Azure network resources.
Now I understand that this is not possible. Can this be provided in the futur as new features ?

Otherwise, can we the GET for the subnets ?

Thank you in advance.

**Reply by DCMattyG** [reply]
I can *DEFINITELY* work on the GET verb for subnets for you. As far as adding in some functionality to IPAM where you'd like to see direct manipulation of Azure resource from within the IPAM tool, perhaps we can talk through in more detail as to the business use cases you're interested in, then create a plan from there.

**Reply by techch01** [reply]
Thank you @DCMattyG for your feedback and for taking the task about GET for subnets level 👍 

We plan to communicate with IPAM from AzureDevops CICD to:
-  Get available IP range
- Then create Vnets/subnets for our Azure resources (like Azure Databricks, etc).

At this stage, I think we will continue to create Vnets/subnets resource from AzureDevops.

One more question, Do you have an idea to authenticate to REST API apart bearer token ?

Thank you in advance.

**Reply by DCMattyG** [reply]
Hi @techch01, we leverage the pre-existing Azure tokens (which we accept as a Bearer token in IPAM) so that we aren't doing authentication separately from Azure (e.g. you don't need to manage authenticating to IPAM and THEN authenticating to Azure separately), and this also keep IPAM out of the business of storing secrets.

What we do see are customers using something like a Service Principal or Managed Identity, and in their Pipeline this SP/MI will auth to IPAM, create a vNET request, get the new CIDR range and Tag, then insert this into the Pipeline for creating a new vNET (using the CIDR and Tag from IPAM). If you're interested in more details on how exactly this would look, I'm sure that's something we can document a bit better for you.

Let us know!

**Reply by techch01** [reply]
Hi @DCMattyG,
Thank you for your return.

Yes, it will be nice to have more documentation about the process described above.

Thanks in advance !

**Reply by DCMattyG** [reply]
Sounds good, @hbendana and I will work on something for you here!

**Reply by techch01** [reply]
Hi @DCMattyG,

Any update from your side about the subject discussed above ?
Thanks in advance !

Best regards,

**Reply by DCMattyG** [reply]
Hey there @techch01, I just wanted you to know that I haven't forgotten about you. We've been working on a few code updates, but now that those are complete I can focus a bit more on this ask. We'll get something put together soon for your review. Thanks!
